FreeRTOS_TCP_IP.c

```diff
@@ -3841,48 +3854,14 @@
             }
         #endif /* ipconfigUSE_CALLBACKS */
 
-        #if ( ipconfigSUPPORT_SELECT_FUNCTION == 1 )
-            {
-                /* Child socket of listening sockets will inherit the Socket Set
-                 * Otherwise the owner has no chance of including it into the set. */
-                if( pxSocket->pxSocketSet != NULL )
-                {
-                    pxNewSocket->pxSocketSet = pxSocket->pxSocketSet;
-                    pxNewSocket->xSelectBits = pxSocket->xSelectBits | ( ( EventBits_t ) eSELECT_READ ) | ( ( EventBits_t ) eSELECT_EXCEPT );
-                }
-            }
-        #endif /* ipconfigSUPPORT_SELECT_FUNCTION */
-
         /* And bind it to the same local port as its parent. */
         xAddress.sin_addr = *ipLOCAL_IP_ADDRESS_POINTER;
         xAddress.sin_port = FreeRTOS_htons( pxSocket->usLocalPort );
 
-        #if ( ipconfigTCP_HANG_PROTECTION == 1 )
-            {
-                /* Only when there is anti-hanging protection, a socket may become an
-                 * orphan temporarily.  Once this socket is really connected, the owner of
-                 * the server socket will be notified. */
-
-                /* When bPassQueued is true, the socket is an orphan until it gets
-                 * connected. */
-                pxNewSocket->u.xTCP.bits.bPassQueued = pdTRUE_UNSIGNED;
-                pxNewSocket->u.xTCP.pxPeerSocket = pxSocket;
-            }
-        #else
-            {
-                /* A reference to the new socket may be stored and the socket is marked
-                 * as 'passable'. */
-
-                /* When bPassAccept is true, this socket may be returned in a call to
-                 * accept(). */
-                pxNewSocket->u.xTCP.bits.bPassAccept = pdTRUE_UNSIGNED;
-
-                if( pxSocket->u.xTCP.pxPeerSocket == NULL )
-                {
-                    pxSocket->u.xTCP.pxPeerSocket = pxNewSocket;
-                }
-            }
-        #endif /* if ( ipconfigTCP_HANG_PROTECTION == 1 ) */
+        /* When bPassQueued is true, the socket is an orphan until it gets
+         * connected. */
+        pxNewSocket->u.xTCP.bits.bPassQueued = pdTRUE_UNSIGNED;
+        pxNewSocket->u.xTCP.pxPeerSocket = pxSocket;
```

Comment:
Okay, but maybe this remains as in independent issue.

Reply:
@bernd-edlinger wrote:

> Also remove the code in `prvTCPSocketCopy()` which makes the
> child socket inherit the socket set of the parent.
> This is most likely just a work-around that makes the select
> wake up when the first user data is received.

I am sure you have explained it earlier, but what was the reason of these changes?
Did you observe a race between the IP-task and the application using select?

Reply:
From the beginning I was wondering where/how the select is supposed to wake up when
`bPassAccept` is set to true here (the case w/o hang protection).
I see the only code path which does this in `vTCPStateChange()`
https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/d425fbd70cb9459f4b9f722668be8cf6de42f75d/FreeRTOS_TCP_IP.c#L1754-L1759
which is guarded by https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/d425fbd70cb9459f4b9f722668be8cf6de42f75d/FreeRTOS_TCP_IP.c#L1730
My other concern, is, when a socket that is not yet in the eESTABLISHED is allowed
to be returned from `FreeRTOS_accept()` would it work as expected or would it not.
So I tried to setup a test for this, I have a server which uses select to wait for connections,
and usually the protocol requires the client to send something first.
So I was surprised that the select did only wake up, after the first bytes arrived,
but if I change the protocol make the server send a "hello" first, the protocol gets stuck.